### PR TITLE
Savings module deposits

### PR DESCRIPTION
**Pull request by denalimarsh**
Implements deposits for the savings module.

Base branch should be changed to master before merge.

### CODE REVIEW

File: x/savings/keeper/deposit.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		return err
+	}
+
+	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, depositor, types.ModuleAccountName, coins)
+	if err != nil {
+		return err
+	}
+
+	currDeposit, foundDeposit := k.GetDeposit(ctx, depositor)
+	var amount sdk.Coins
```
Comment (karzak) [inline x/savings/keeper/deposit.go:23]:
```suggestion
	amount := coins
```

File: x/savings/keeper/deposit.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	if foundDeposit {
+		amount = currDeposit.Amount.Add(coins...)
+	} else {
```
Comment (karzak) [inline x/savings/keeper/deposit.go:26]: can ignore this else if you initialize as `amount := coins`

File: x/savings/keeper/params.go [hunk 3]
```diff
@@ -17,3 +17,16 @@ func (k Keeper) GetParams(ctx sdk.Context) types.Params {
 func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
 	k.paramSubspace.SetParamSet(ctx, &params)
 }
+
+// IsDenomSupported returns a boolean indicating if a denom is supported
+func (k Keeper) IsDenomSupported(ctx sdk.Context, denom string) bool {
+	var p types.Params
+	k.paramSubspace.GetParamSet(ctx, &p)
```
Comment (karzak) [inline x/savings/keeper/params.go:24]:
```suggestion
	params := k.GetParams(ctx)
```

**Review by karzak — CHANGES_REQUESTED** [review]
First pass, will give another pass to the tests and do some manual testing tomorrow.

File: x/savings/keeper/deposit.go (hunk 2 above)
Comment (denalimarsh) [reply x/savings/keeper/deposit.go]: Updated in https://github.com/Kava-Labs/kava/pull/1192/commits/153393d4a9ba56d21627f8f597a2da860a2faf3a

File: x/savings/keeper/deposit_test.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	tmtime "github.com/tendermint/tendermint/types/time"
+
+	"github.com/kava-labs/kava/app"
+	"github.com/kava-labs/kava/x/savings/types"
+)
+
+func (suite *KeeperTestSuite) TestDeposit() {
+	type args struct {
+		depositor                 sdk.AccAddress
+		amount                    sdk.Coins
```
Comment (karzak) [inline x/savings/keeper/deposit_test.go:18]: would call this `depositAmount` and add `initialDepositorBalance` as an arg (see below)

File: x/savings/keeper/deposit_test.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+			// create new app with one funded account
+
+			// Initialize test app and set context
+			tApp := app.NewTestApp()
+			ctx := tApp.NewContext(true, tmproto.Header{Height: 1, Time: tmtime.Now()})
+			authGS := app.NewFundedGenStateWithCoins(
+				tApp.AppCodec(),
+				[]sdk.Coins{
+					sdk.NewCoins(
+						sdk.NewCoin("bnb", sdk.NewInt(1000)),
```
Comment (karzak) [inline x/savings/keeper/deposit_test.go:106]: I would set the initial depositor balance as part of the `args` ie `tc.args.initialDepositorBalance`

File: x/savings/keeper/deposit_test.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+						sdk.NewCoin("btcb", sdk.NewInt(1000)),
+					),
+				},
+				[]sdk.AccAddress{tc.args.depositor},
+			)
+			savingsGS := types.NewGenesisState(types.NewParams(
+				[]string{"bnb", "btcb", "ukava"},
```
Comment (karzak) [inline x/savings/keeper/deposit_test.go:113]: would move the allowed denoms to `args` as well

**Review by karzak — CHANGES_REQUESTED** [review]
Looking good, few changes to test structure for the main deposits test.

File: x/savings/keeper/deposit.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+	return nil
+}
+
+// ValidateDeposit validates a deposit
+func (k Keeper) ValidateDeposit(ctx sdk.Context, coins sdk.Coins) error {
+	for _, coin := range coins {
+		supported := k.IsDenomSupported(ctx, coin.Denom)
+		if supported == false {
+			return types.ErrInvalidDepositDenom
```
Comment (karzak) [inline x/savings/keeper/deposit.go:46]: I would wrap this error with the denom that caused it - currently the error is just

`raw_log: 'failed to execute message; message index: 0: invalid deposit denom'`

Would like:

`raw_log: 'failed to execute message; message index: 0: invalid deposit denom: bnb'`

**Review by karzak — CHANGES_REQUESTED** [review]

**Comment by karzak** [comment]
Manual testing with kvtool complete - LGTM after the above requested changes are added.

File: x/savings/keeper/deposit.go (hunk 7 above)
Comment (denalimarsh) [reply x/savings/keeper/deposit.go]: Added in https://github.com/Kava-Labs/kava/pull/1192/commits/b2cbe147ae733740910887e81c5090ece674c7f6

File: x/savings/keeper/deposit_test.go (hunk 6 above)
Comment (denalimarsh) [reply x/savings/keeper/deposit_test.go]: Updated in latest

File: x/savings/keeper/deposit_test.go (hunk 5 above)
Comment (denalimarsh) [reply x/savings/keeper/deposit_test.go]: Updated in latest

File: x/savings/keeper/deposit_test.go (hunk 4 above)
Comment (denalimarsh) [reply x/savings/keeper/deposit_test.go]: Updated in latest

**Pull request merged**